File: packages/sdk-sidecar/test/encryption.test.ts
import { randomUUID } from "node:crypto";
import { Metadata, status, type ServiceError } from "@grpc/grpc-js";
import type { EncryptInput, EncryptParams, EncryptResult } from "@zama-fhe/sdk";
import { bytesToHex, getAddress } from "viem";
import { expect, test, vi } from "vitest";
import { createEncryptionValidationBackend } from "../../sdk/src/test-fixtures/encryption.js";
import { bytes } from "../src/encoding.js";
import { serviceError } from "../src/errors.js";
import { encrypt as adaptEncryption, encryptInput } from "../src/encryption.js";
import type {
  EncryptInput as WireInput,
  EncryptRequest,
  EncryptResponse,
} from "../src/generated/zama/sdk/v1alpha1/sidecar.js";
import type { ContextSdk } from "../src/runtime.js";
import {
  encryptionFixture,
  encryptionServer,
  type Scenario,
  scenarioUrl,
} from "./support/encryption.js";
import { createContext } from "./support/harness.js";

const contractAddress = getAddress("0x1234567890123456789012345678901234567890");
const userAddress = getAddress("0xabcdefabcdefabcdefabcdefabcdefabcdefabcd");
const inputs: EncryptInput[] = [
  ...([8, 16, 32, 64, 128, 256] as const).map((bits): EncryptInput => ({
    type: `euint${bits}`,
    value: (1n << BigInt(bits)) - 1n,
  })),
  { type: "ebool", value: false },
  { type: "ebool", value: true },
  { type: "ebool", value: 0n },
  { type: "ebool", value: 1n },
  { type: "eaddress", value: userAddress },
];
function wire(input: EncryptInput): WireInput {
  switch (input.type) {
    case "ebool":
      return typeof input.value === "boolean"
        ? { value: { $case: "ebool", ebool: input.value } }
        : { value: { $case: "eboolBigint", eboolBigint: input.value.toString() } };
    case "euint8":
      return { value: { $case: "euint8", euint8: input.value.toString() } };
    case "euint16":
      return { value: { $case: "euint16", euint16: input.value.toString() } };
    case "euint32":
      return { value: { $case: "euint32", euint32: input.value.toString() } };
    case "euint64":
      return { value: { $case: "euint64", euint64: input.value.toString() } };
    case "euint128":
      return { value: { $case: "euint128", euint128: input.value.toString() } };
    case "euint256":
      return { value: { $case: "euint256", euint256: input.value.toString() } };
    case "eaddress":
    default:
      return { value: { $case: "eaddress", eaddress: bytes(input.value) } };
  }
}
async function harness(scenario: Scenario = "", signerEnabled = false) {
  const server = await encryptionServer();
  const contextId = await createContext(server.client, {
    config: {
      chains: [
        { id: 31337n, relayerUrl: scenarioUrl(scenario), auth: undefined, provider: undefined },
      ],
      processRuntime: undefined,
      relayers: undefined,
    },
    signerEnabled,
    account: signerEnabled ? { address: bytes(contractAddress), chainId: 31337n } : undefined,
  });
  const request = (values = inputs, timeoutMs?: number): EncryptRequest => ({
    operation: { contextId, operationId: randomUUID() },
    values: values.map(wire),
    contractAddress: bytes(contractAddress),
    userAddress: bytes(userAddress),
    timeoutMs,
  });
  const encrypt = (value = request()) =>
    new Promise<EncryptResponse>((resolve, reject) =>
      server.client.encrypt(value, (error, response) =>
        error ? reject(error) : resolve(response),
      ),
    );
  return { ...server, request, encrypt };
}
function proof(result: EncryptResult) {
  return JSON.parse(Buffer.from(result.inputProof.slice(2), "hex").toString());
}
function decode(result: EncryptResponse): EncryptResult {
  return {
    encryptedValues: result.encryptedValues.map((value) => bytesToHex(value)),
    inputProof: bytesToHex(result.inputProof),
  };
}

test.each([undefined, 0, 1234, 0xffff_ffff])(
  "direct SDK and wire preserve all types, binding, and timeout %s",
  async (timeout) => {
    const direct = encryptionFixture("", undefined);
    const remote = await harness();
    try {
      const params: EncryptParams = { values: inputs, contractAddress, userAddress };
      const options = timeout === undefined ? undefined : { timeout };
      const local = await direct.sdk.encrypt(params, options);
      const result = decode(await remote.encrypt(remote.request(inputs, timeout)));
      expect(proof(result)).toEqual(proof(local));
      expect(result.encryptedValues).toHaveLength(inputs.length);
      expect(result.encryptedValues.every((value) => /^0x[0-9a-f]{64}$/.test(value))).toBe(true);
      expect(result.encryptedValues).not.toEqual(local.encryptedValues);
      const received = remote.fixtures[0]!.encryptValues.mock.calls[0]![0];
      expect(received.values).toEqual(inputs);
      expect(received.options).toHaveProperty("signal");
      expect(Object.hasOwn(received.options!, "timeout")).toBe(timeout !== undefined);
    } finally {
      direct.sdk.dispose();
      await remote.close();
    }
  },
);

test.each([-1, 1.5, 0x1_0000_0000])(
  "the direct adapter rejects invalid timeout %s before calling the SDK",
  async (timeoutMs) => {
    const sdk = { encrypt: vi.fn() } as unknown as ContextSdk;
    await expect(
      adaptEncryption(
        sdk,
        {
          operation: undefined,
          values: inputs.map(wire),
          contractAddress: bytes(contractAddress),
          userAddress: bytes(userAddress),
          timeoutMs,
        },
        new AbortController().signal,
      ),
    ).rejects.toMatchObject({
      code: "INVALID_ARGUMENT",
      message: "Timeout must be an unsigned 32-bit integer.",
    });
    expect(sdk.encrypt).not.toHaveBeenCalled();
  },
);

test("empty inputs and out-of-range integers reach SDK unchanged", async () => {
  const remote = await harness();
  try {
    for (const values of [
      [],
      [
        { type: "euint8", value: -1n },
        { type: "euint8", value: 256n },
      ] satisfies EncryptInput[],
    ]) {
      const result = await remote.encrypt(remote.request(values));
      expect(result.encryptedValues).toHaveLength(values.length);
      expect(remote.fixtures[0]!.encryptValues.mock.lastCall![0].values).toEqual(values);
    }
  } finally {
    await remote.close();
  }
});

test.each<{ scenario: Scenario; details: string; retryable: string[]; retryAfter: string[] }>([
  {
    scenario: "rate-limited",
    details: "Encryption service busy",
    retryable: ["true"],
    retryAfter: ["7"],
  },
  {
    scenario: "fractional-retry-hint",
    details: "Encryption retry hint is fractional",
    retryable: ["true"],
    retryAfter: [],
  },
  {
    scenario: "unavailable",
    details: "Encryption service unavailable",
    retryable: ["false"],
    retryAfter: [],
  },
])(
  "SDK failure preserves only valid whole-second retry details for $scenario",
  async ({ scenario, details, retryable, retryAfter }) => {
    const direct = encryptionFixture(scenario, undefined);
    const remote = await harness(scenario);
    try {
      const expected = await direct.sdk
        .encrypt({ values: inputs, contractAddress, userAddress })
        .catch(serviceError);
      const actual = await remote.encrypt().catch((error: ServiceError) => error);
      expect(actual).toMatchObject({ code: (expected as ServiceError).code, details });
      for (const key of [
        "zama-error-code",
        "zama-error-retryable",
        "zama-error-retry-after-seconds",
      ]) {
        expect((actual as ServiceError).metadata.get(key)).toEqual(
          (expected as ServiceError).metadata.get(key),
        );
      }
      expect((actual as ServiceError).metadata.get("zama-error-retryable")).toEqual(retryable);
      expect((actual as ServiceError).metadata.get("zama-error-retry-after-seconds")).toEqual(
        retryAfter,
      );
    } finally {
      direct.sdk.dispose();
      await remote.close();
    }
  },
);

test("concurrent public encryption and cancellation remain isolated", async () => {
  const remote = await harness("cancelled", true);
  try {
    const start = () => {
      let call: ReturnType<typeof remote.client.encrypt>;
      const result = new Promise<ServiceError>((resolve, reject) => {
        call = remote.client.encrypt(remote.request(), (error) => {
          if (error) {
            resolve(error);
          } else {
            reject(new Error("blocked encryption unexpectedly completed"));
          }
        });
      });
      return { call: call!, result };
    };
    const first = start();
    const second = start();
    await expect.poll(() => remote.fixtures[0]!.encryptValues.mock.calls.length).toBe(2);
    const firstSignal = remote.fixtures[0]!.encryptValues.mock.calls[0]![0].options!.signal!;
    const secondSignal = remote.fixtures[0]!.encryptValues.mock.calls[1]![0].options!.signal!;

    first.call.cancel();
    expect(await first.result).toMatchObject({ code: status.CANCELLED });
    await expect.poll(() => firstSignal.aborted).toBe(true);
    expect(secondSignal.aborted).toBe(false);

    second.call.cancel();
    expect(await second.result).toMatchObject({ code: status.CANCELLED });
    await expect.poll(() => secondSignal.aborted).toBe(true);
  } finally {
    await remote.close();
  }
});

test.each(["cancel", "deadline", "close"])("%s aborts SDK encryption", async (mode) => {
  const remote = await harness("cancelled");
  try {
    const request = remote.request();
    let call: ReturnType<typeof remote.client.encrypt>;
    const result = new Promise<EncryptResponse>((resolve, reject) => {
      call = remote.client.encrypt(
        request,
        new Metadata(),
        mode === "deadline" ? { deadline: Date.now() + 100 } : {},
        (error, response) => (error ? reject(error) : resolve(response)),
      );
    });
    const failed = result.catch((error: ServiceError) => error);
    await expect.poll(() => remote.fixtures[0]!.encryptValues.mock.calls.length).toBe(1);
    const signal = remote.fixtures[0]!.encryptValues.mock.calls[0]![0].options!.signal!;
    if (mode === "cancel") {
      call!.cancel();
    }
    if (mode === "close") {
      await new Promise<void>((resolve, reject) =>
        remote.client.closeContext({ contextId: request.operation!.contextId }, (error) =>
          error ? reject(error) : resolve(),
        ),
      );
    }
    expect(await failed).toMatchObject({
      code: mode === "deadline" ? status.DEADLINE_EXCEEDED : status.CANCELLED,
    });
    await expect.poll(() => signal.aborted).toBe(true);
  } finally {
    await remote.close();
  }
});

test.each(["", "01", "-0", "1.5", "0x10", "+1", " 1"])(
  "rejects malformed decimal %j at the wire boundary",
  (euint64) => {
    expect(() => encryptInput({ value: { $case: "euint64", euint64 } })).toThrow(
      "canonical decimal",
    );
  },
);
test("rejects untyped values and invalid address lengths", async () => {
  expect(() => encryptInput({ value: undefined })).toThrow("requires a typed value");
  for (const length of [0, 19, 21]) {
    expect(() =>
      encryptInput({ value: { $case: "eaddress", eaddress: Buffer.alloc(length) } }),
    ).toThrow("20 bytes");
  }
  const remote = await harness();
  try {
    for (const addresses of [
      { userAddress: Buffer.alloc(19) },
      { contractAddress: Buffer.alloc(0) },
    ]) {
      await expect(remote.encrypt({ ...remote.request(), ...addresses })).rejects.toMatchObject({
        code: status.INVALID_ARGUMENT,
      });
    }
    expect(remote.fixtures[0]!.encryptValues).not.toHaveBeenCalled();
  } finally {
    await remote.close();
  }
});

test("canonical backend numeric rejection errors match direct SDK errors", async () => {
  const validate = createEncryptionValidationBackend();
  const direct = encryptionFixture("invalid-input", undefined);
  const remote = await harness("invalid-input");
  try {
    for (const values of [
      [{ type: "euint8", value: -1n }],
      [{ type: "euint8", value: 256n }],
      [{ type: "euint256", value: 1n << 256n }],
      [{ type: "ebool", value: -1n }],
      [{ type: "ebool", value: 2n }],
    ] as EncryptInput[][]) {
      const params = { values, contractAddress, userAddress };
      const backendError = await validate(params).catch((error: Error) => error);
      expect(backendError).toBeInstanceOf(Error);
      expect((backendError as Error).message).not.toContain("network access");
      const local = await direct.sdk.encrypt(params).catch(serviceError);
      const result = await remote
        .encrypt(remote.request(values))
        .catch((error: ServiceError) => error);
      expect(result).toMatchObject({
        code: (local as ServiceError).code,
        details: (local as ServiceError).details,
      });
      expect((result as ServiceError).metadata.get("zama-error-code")).toEqual([
        "ENCRYPTION_FAILED",
      ]);
      expect((result as ServiceError).metadata.get("zama-error-retryable")).toEqual(["false"]);
    }
  } finally {
    direct.sdk.dispose();
    await remote.close();
  }
});

test("explicit user binding stays independent from the connected wallet", async () => {
  const remote = await harness("", true);
  try {
    const result = decode(await remote.encrypt());
    expect(proof(result).userAddress).toBe(userAddress);
    const changed = await remote.encrypt({
      ...remote.request(),
      contractAddress: bytes(userAddress),
      userAddress: bytes(contractAddress),
    });
    expect(proof(decode(changed))).toMatchObject({
      contractAddress: userAddress,
      userAddress: contractAddress,
    });
  } finally {
    await remote.close();
  }
});

test("a direct SDK call observes the same abort signal behavior", async () => {
  const direct = encryptionFixture("cancelled", undefined);
  const controller = new AbortController();
  try {
    const result = direct.sdk.encrypt(
      { values: inputs, contractAddress, userAddress },
      { signal: controller.signal },
    );
    const failed = result.catch((error: Error) => error);
    await expect.poll(() => direct.encryptValues.mock.calls.length).toBe(1);
    controller.abort();
    expect(await failed).toMatchObject({ code: "ENCRYPTION_FAILED" });
    expect(direct.encryptValues.mock.lastCall![0].options!.signal!.aborted).toBe(true);
  } finally {
    direct.sdk.dispose();
  }
});

File: clients/go/encryption_test.go
package sidecar

import (
	"bytes"
	"context"
	"errors"
	"math/big"
	"testing"
	"time"

	"github.com/ethereum/go-ethereum/common"
	pb "github.com/zama-ai/sdk/clients/go/gen/zama/sdk/v1alpha1"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"
)

func TestEncryptWire(t *testing.T) {
	requests := make(chan *pb.EncryptRequest, 4)
	handle := bytes.Repeat([]byte{0xab}, 32)
	client := testClient(t, &pb.UnimplementedSidecarServiceServer{}, func(_ context.Context, request any, _ *grpc.UnaryServerInfo, _ grpc.UnaryHandler) (any, error) {
		if req, ok := request.(*pb.EncryptRequest); ok {
			requests <- req
			handles := make([][]byte, len(req.Values))
			for i := range handles {
				handles[i] = handle
			}
			return &pb.EncryptResponse{EncryptedValues: handles, InputProof: []byte{0, 255, 1}}, nil
		}
		return &pb.CreateContextResponse{ContextId: "encrypt"}, nil
	})
	sdk := unsignedSDK(t, client)
	large := new(big.Int).Lsh(big.NewInt(1), 255)
	params := EncryptParams{ContractAddress: common.HexToAddress("0x1111111111111111111111111111111111111111"), UserAddress: common.HexToAddress("0x2222222222222222222222222222222222222222")}
	params.Values = []EncryptInput{
		Euint8(large), Euint16(large), Euint32(large), Euint64(large), Euint128(large), Euint256(large),
		Ebool(false), Ebool(true), EboolBigInt(big.NewInt(0)), EboolBigInt(big.NewInt(1)),
		Eaddress(params.UserAddress), Euint8(big.NewInt(-1)),
	}
	result, err := sdk.Encrypt(testContext(t), params, EncryptOptions{})
	if err != nil {
		t.Fatal(err)
	}
	req := <-requests
	if req.TimeoutMs != nil || req.Operation.ContextId != "encrypt" || req.Operation.OperationId == "" || !bytes.Equal(req.ContractAddress, params.ContractAddress.Bytes()) || !bytes.Equal(req.UserAddress, params.UserAddress.Bytes()) {
		t.Fatalf("parameters lost: %v", req)
	}
	integers := []string{req.Values[0].GetEuint8(), req.Values[1].GetEuint16(), req.Values[2].GetEuint32(), req.Values[3].GetEuint64(), req.Values[4].GetEuint128(), req.Values[5].GetEuint256()}
	for _, value := range integers {
		if value != large.String() {
			t.Fatal("integer lost precision/type")
		}
	}
	if _, ok := req.Values[6].Value.(*pb.EncryptInput_Ebool); !ok || req.Values[6].GetEbool() || !req.Values[7].GetEbool() {
		t.Fatal("boolean representation lost")
	}
	if req.Values[8].GetEboolBigint() != "0" || req.Values[9].GetEboolBigint() != "1" || req.Values[11].GetEuint8() != "-1" || !bytes.Equal(req.Values[10].GetEaddress(), params.UserAddress.Bytes()) {
		t.Fatal("input representation lost")
	}
	if len(result.EncryptedValues) != len(params.Values) || result.EncryptedValues[0] != common.BytesToHash(handle) || !bytes.Equal(result.InputProof, []byte{0, 255, 1}) {
		t.Fatal("result changed")
	}
	for _, timeout := range []uint32{0, 500} {
		if _, err := sdk.Encrypt(testContext(t), EncryptParams{}, EncryptOptions{TimeoutMS: &timeout}); err != nil {
			t.Fatal(err)
		}
		req = <-requests
		if req.TimeoutMs == nil || *req.TimeoutMs != timeout || len(req.Values) != 0 {
			t.Fatalf("explicit timeout or empty inputs changed: %v", req)
		}
	}
}

func TestEncryptErrorsAndCancellation(t *testing.T) {
	started, cancelled := make(chan struct{}, 1), make(chan struct{}, 1)
	client := testClient(t, &pb.UnimplementedSidecarServiceServer{}, func(ctx context.Context, request any, _ *grpc.UnaryServerInfo, _ grpc.UnaryHandler) (any, error) {
		if req, ok := request.(*pb.EncryptRequest); ok {
			if req.TimeoutMs != nil {
				grpc.SetTrailer(ctx, metadata.Pairs("zama-error-code", "RELAYER_REQUEST_FAILED", "zama-error-retryable", "true", "zama-error-retry-after-seconds", "7"))
				return nil, status.Error(codes.Unavailable, "fixture failure")
			}
			started <- struct{}{}
			<-ctx.Done()
			cancelled <- struct{}{}
			return nil, ctx.Err()
		}
		return &pb.CreateContextResponse{ContextId: "encrypt"}, nil
	})
	sdk := unsignedSDK(t, client)
	timeout := uint32(13)
	_, err := sdk.Encrypt(testContext(t), EncryptParams{}, EncryptOptions{TimeoutMS: &timeout})
	var rpc *RPCError
	if !errors.As(err, &rpc) || rpc.Code != "RELAYER_REQUEST_FAILED" || !rpc.Retryable || rpc.RetryAfterSeconds == nil || *rpc.RetryAfterSeconds != 7 || status.Code(err) != codes.Unavailable {
		t.Fatalf("structured error lost: %#v", err)
	}
	ctx, cancel := context.WithCancel(testContext(t))
	defer cancel()
	done := make(chan error, 1)
	go func() { _, err := sdk.Encrypt(ctx, EncryptParams{}, EncryptOptions{}); done <- err }()
	select {
	case <-started:
	case <-time.After(time.Second):
		t.Fatal("request did not start")
	}
	cancel()
	if err := <-done; status.Code(err) != codes.Canceled {
		t.Fatalf("unexpected cancellation: %v", err)
	}
	select {
	case <-cancelled:
	case <-time.After(time.Second):
		t.Fatal("server not cancelled")
	}
}

func TestEncryptRejectsMalformedHandle(t *testing.T) {
	client := testClient(t, &pb.UnimplementedSidecarServiceServer{}, func(_ context.Context, request any, _ *grpc.UnaryServerInfo, _ grpc.UnaryHandler) (any, error) {
		if _, ok := request.(*pb.EncryptRequest); ok {
			return &pb.EncryptResponse{EncryptedValues: [][]byte{{1}}}, nil
		}
		return &pb.CreateContextResponse{ContextId: "encrypt"}, nil
	})
	if _, err := unsignedSDK(t, client).Encrypt(testContext(t), EncryptParams{Values: []EncryptInput{Ebool(true)}}, EncryptOptions{}); err == nil {
		t.Fatal("malformed handle accepted")
	}
}

func TestEncryptRejectsHandleCountMismatch(t *testing.T) {
	handle := bytes.Repeat([]byte{0xab}, 32)
	client := testClient(t, &pb.UnimplementedSidecarServiceServer{}, func(_ context.Context, request any, _ *grpc.UnaryServerInfo, _ grpc.UnaryHandler) (any, error) {
		if _, ok := request.(*pb.EncryptRequest); ok {
			return &pb.EncryptResponse{EncryptedValues: [][]byte{handle}}, nil
		}
		return &pb.CreateContextResponse{ContextId: "encrypt"}, nil
	})
	params := EncryptParams{Values: []EncryptInput{Ebool(true), Ebool(false)}}
	if _, err := unsignedSDK(t, client).Encrypt(testContext(t), params, EncryptOptions{}); err == nil {
		t.Fatal("handle count mismatch accepted")
	}
}

func TestEncryptRejectsAbsentInputs(t *testing.T) {
	for _, absent := range []struct {
		name    string
		input   EncryptInput
		message string
	}{
		{"zero value", EncryptInput{}, "encryption input 1: missing value"},
		{"nil integer", Euint64(nil), "encryption input 1: euint64 value is nil"},
		{"nil boolean integer", EboolBigInt(nil), "encryption input 1: ebool value is nil"},
	} {
		t.Run(absent.name, func(t *testing.T) {
			params := EncryptParams{Values: []EncryptInput{Ebool(true), absent.input}}
			_, err := (&SDKContext{}).Encrypt(context.Background(), params, EncryptOptions{})
			if err == nil || err.Error() != absent.message {
				t.Fatalf("absent input accepted: %v", err)
			}
		})
	}
}
